File: matlab/src/lang/constant/SignalType.m
classdef (Sealed) SignalType
    % Enumeration of signal types for signal classes.
    %
    % Copyright (c) 2012-2026, Imperial College London
    % All rights reserved.

    properties (Constant)
        NEGATIVE = 1;     % Negative customer signal (removes jobs)
        REPLY = 0;        % Reply signal (triggers unblocking)
        CATASTROPHE = 2;  % Catastrophe signal (removes ALL jobs)
    end

    methods (Static)
        function text = toText(type)
            % TEXT = TOTEXT(TYPE)
            switch type
                case SignalType.NEGATIVE
                    text = 'negative';
                case SignalType.REPLY
                    text = 'reply';
                case SignalType.CATASTROPHE
                    text = 'catastrophe';
                otherwise
                    line_error(mfilename, 'Unrecognized signal type.');
            end
        end

        function type = fromText(text)
            % TYPE = FROMTEXT(TEXT)
            if iscell(text)
                text = text{:};
            end
            switch lower(text)
                case 'negative'
                    type = SignalType.NEGATIVE;
                case 'reply'
                    type = SignalType.REPLY;
                case 'catastrophe'
                    type = SignalType.CATASTROPHE;
                otherwise
                    line_error(mfilename, 'Unrecognized signal type text.');
            end
        end
    end
end

File: jar/src/main/java/jline/lang/constant/SignalType.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */

package jline.lang.constant;

import java.io.Serializable;

/**
 * Enumeration of signal types for signal classes in queueing networks.
 */
public enum SignalType implements Serializable {
    NEGATIVE(1),
    REPLY(0),
    CATASTROPHE(2);

    private final int id;

    SignalType(int id) {
        this.id = id;
    }

    public int getID() {
        return id;
    }

    public static SignalType fromID(int id) {
        for (SignalType type : values()) {
            if (type.id == id) {
                return type;
            }
        }
        return null;
    }

    public static String toText(SignalType type) {
        if (type == null) {
            return null;
        }
        switch (type) {
            case NEGATIVE:
                return "negative";
            case REPLY:
                return "reply";
            case CATASTROPHE:
                return "catastrophe";
            default:
                return type.name();
        }
    }

    public static SignalType fromText(String text) {
        if (text == null || text.isEmpty()) {
            return null;
        }
        switch (text.toLowerCase()) {
            case "negative":
                return NEGATIVE;
            case "reply":
                return REPLY;
            case "catastrophe":
                return CATASTROPHE;
            default:
                return null;
        }
    }
}
